### PR TITLE
8214158: Implement HostServices.showDocument on macOS without calling AWT

### PR DESCRIPTION
**Pull request by mipastgt**
Fixes the issue by using the "open" command as discussed before.
See: https://bugs.openjdk.java.net/browse/JDK-8214158

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8214158](https://bugs.openjdk.java.net/browse/JDK-8214158): Implement HostServices.showDocument on macOS without calling AWT

### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/634/head:pull/634` \
`$ git checkout pull/634`

Update a local copy of the PR: \
`$ git checkout pull/634` \
`$ git pull https://git.openjdk.java.net/jfx pull/634/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 634`

View PR using the GUI difftool: \
`$ git pr show -t 634`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/634.diff">https://git.openjdk.java.net/jfx/pull/634.diff</a>

</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
This seems like a good RFE. I have a couple overall questions.

First, calling `open` was originally proposed in the since-abandoned Draft PR javafxports/openjdk-jfx#310. As I wrote there:

> One thing of note is that the AWT code, which we are currently calling, uses a native method to call LSOpenURLsWithRole rather than "exec"-ing a new process from Java. I'm not sure why (maybe the open command wasn't as robust when it was implemented? Or maybe there is some benefit to using LSOpenURLsWithRole?), but we will need to make sure that there isn't a regression in behavior as a result."

Can you comment on this?

Second, can you provide a manual test for this? I was thinking something that would go somewhere under `tests/manual` and when run, provide an option to open one of two or three different file types (e.g., a `.txt` file and a `.html` file) that are also checked in to the same dir as the test program. It could use `FIle::toURI` or similar to construct the URI string.

**Comment by mipastgt** [comment]
First of all. Shall I follow the bot request above and create a new pull request? Wouldn't we loose the discussion here then?

**Comment by mipastgt** [comment]
To answer your questions:
1. I have no idea about LSOpenURLsWithRole. I am not a Mac developer.
2. I have manually tested the change:

        Button showHtml = new Button("HTML");
        showHtml.setOnAction(e -> {this.getHostServices().showDocument(new File("data/test.html").toURI().toString());});
        
        Button showTxt = new Button("TXT");
        showTxt.setOnAction(e -> {this.getHostServices().showDocument(new File("data/test.txt").toURI().toString());});
        
        Button showCsv = new Button("CSV");
        showCsv.setOnAction(e -> {this.getHostServices().showDocument(new File("data/test.csv").toURI().toString());});

The code fragment above reliably opens the HTML document in Firefox (my default browser), the text document in BBEdit (my default text editor) and the CSV document in LibreOffice (my default Excel :-)

This also works if you just use the plain strings without conversion to an URI.

Here are the printed Java/FX version strings for completness:

    java.runtime.version: 18-ea+16-926
    javafx.runtime.version: 18-internal+0-2021-09-28-174132

**Comment by kevinrushforth** [comment]
> First of all. Shall I follow the bot request above and create a new pull request? Wouldn't we loose the discussion here then?

As long as you don't otherwise need to update your `master` branch, you can wait to reset it until after this PR is integrated. If you prefer to create a new branch and a new PR, then you could link to the discussion. I'll leave it up to you.

> I have no idea about LSOpenURLsWithRole. I am not a Mac developer.

I think calling `Runtime::exec("open", uri)` is fine. It matches the behavior on Windows and Linux. I'll do some testing, but I don't anticipate any problems.

> I have manually tested the change.

In that case, can you create a new manual test as part of this PR? I recommend a new directory: `tests/manual/desktop/HostServices` (my thinking is that `tests/manual/desktop` could eventually include tests for `FileChooser` or other similar manual tests). I'm open to other names as well. You can look at `tests/manual/web/InputTypeAcceptAttributeTest.java` for an example of a good manual test. I recommend adding at least a `.txt` and a `.html` file. You can hard-code the filenames so you don't need to use a `FileChooser`.

File: tests/manual/desktop/HostServices/test.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+TXT Test
+Test page for HostServices.showDocument.
```
Comment (kevinrushforth) [inline tests/manual/desktop/HostServices/test.txt:2]: Minor: It doesn't show up in the UI, but this file doesn't have a newline at the end of file

**Review by kevinrushforth — APPROVED** [review]
Both the fix and the test look good. I left two minor formatting comments, but I'll approve it as-is. If you want to fix them, I'll re-approve.

**Review by kevinrushforth — APPROVED** [review]

**Review by arapte — APPROVED** [review]
Looks good. Verified on Mac and Windows.

**Pull request closed without merging**